### PR TITLE
Make def generated from givens not synthetic

### LINKED ISSUE #12949: Export is missing more refined "given with" instances
When a `given with` expression returns a more refined instance, export is missing it from the implicit scope.
Regular givens or transparent inline givens are OK.

## Compiler version

v3.0.1-RC2 (also tested on nightly)

## Minimized code

```Scala
object Catch22:
  trait TC[V]
  object TC:
    export Hodor.TC.given

object Hodor:
  object TC:
    import Catch22.TC
    given fromString[V <: String]: TC[V] = ???
    transparent inline given fromDouble[V <: Double]: TC[V] =
      new TC[V]:
        type Out = Double
    given fromInt[V <: Int]: TC[V] with
      type Out = Int

object Test:
  summon[Catch22.TC["hi"]] //works
  summon[Catch22.TC[7.7]] //works
  summon[Catch22.TC[1]] //error
```

## Output

```scala
no implicit argument of type Catch22.TC[(1 : Int)] was found for parameter x of method summon in object Predef.
I found:

    Catch22.TC.fromDouble[V]

But given instance fromDouble in object TC does not match type Catch22.TC[(1 : Int)].
```

## Expectation
No error.

### PR DESCRIPTION
**Pull request by odersky**
Fixes #12949

### CODE REVIEW

**Comment by odersky** [comment]
The problem is how to know whether some method is a structural given instance or
an implicit conversion associated with an implicit class. Previously we characterized
these methods by their Synthetic flag, but that meant that they were not automatically
exported. So we now drop Synthetic and use a flag `GivenClass` on the class half of
the pair.

The old scheme is still recognized in order to maintain Tasty backwards compatibility.
Tasty forwards compatibility is lost, that's why we need to wait for the next minor release.

A better fix would be to give the synthetic class of a structural given instance a special semantic name.
That would also get rid of the weirdness that the generated class of a given instance like
```scala
given foo: C with ...
```
is visible under the name `foo`. But that would break backwards Tasty compatibility, so we have to wait 
until we can implement that.

**Comment by soronpo** [comment]
> The old scheme is still recognized in order to maintain Tasty backwards compatibility.
> Tasty forwards compatibility is lost, that's why we need to wait for the next minor release.

So this means this PR will be merged only for 3.1?

**Comment by odersky** [comment]
@soronpo Yes, but I think we'll move to 3.1 soon anyway.

**Comment by soronpo** [comment]
> 
> 
> @soronpo Yes, but I think we'll move to 3.1 soon anyway.

Then I better get to finding more compatibility braking bugs soon :wink:

**Comment by soronpo** [comment]
> 
> 
> @soronpo Yes, but I think we'll move to 3.1 soon anyway.

Then I propose creating a warning message for 3.0.x notifying more refined "given with" types are not supported yet, if such structures are used.

**Comment by odersky** [comment]
@soronpo Not sure exactly what the warning should do and who would design and implement it.

**Comment by soronpo** [comment]
> Not sure exactly what the warning should do 

Currently (until 3.1) `given ... with ... type T` structures are not valid. They exist in the language, but the compiler does not support them fully. This leads to an error that the user needs to figure out, possibly spend several hour minimizing, possibly find this issue, and possibly gets frustrated. Deprecating this feature until 3.1 is the way to go, IMO. I don't mind falling into the pit myself. I know Scala 3 is not yet completely charted territory and I tread lightly. I just want to prevent others from falling into the same pit without even putting a warning sign in front of it. 

>and who would design and implement it.

I don't mind helping out, but unfortunately currently my table is full. I'm trying my best minimizing and documenting the issues I find as I'm migrating my library to Scala 3.

**Comment by odersky** [comment]
> Currently (until 3.1) given ... with ... type T structures are not valid. 

That's a bit extremist. What does not work is wildcard exports for them. They are nevertheless widely used, so introducing a blanket warning would be extremely annoying. If someone can find a targeted warning that is clearly not annoying then I am happy to see a PR. But personally i think it will be very difficult to achieve and not worth the effort.

File: tests/semanticdb/expect/Classes.expect.scala [hunk 1]
```diff
@@ -11,7 +11,7 @@ case class C4/*<-classes::C4#*/(x/*<-classes::C4#x.*/: Int/*->scala::Int#*/)
 object C4/*<-classes::C4.*/
 
 object M/*<-classes::M.*/ {
-  implicit class C5/*<-classes::M.C5#*/(x/*<-classes::M.C5#x.*/: Int/*->scala::Int#*/)
+  /*->classes::M.C5().(x)*/implicit class C5/*<-classes::M.C5#*//*<-classes::M.C5().*/(x/*<-classes::M.C5#x.*//*<-classes::M.C5().(x)*/: Int/*->scala::Int#*/)
```
Comment (tgodzik) [inline tests/semanticdb/expect/Classes.expect.scala:14]: This should not be added to semanticdb occurrences, they are synthetic for the source code point of view :thinking: 

For example ` /*->classes::M.C5().(x)*/` doesn't really correspond to any symbol, it's at the start of the line, where no symbol is declared.

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 2]
```diff
@@ -809,7 +809,10 @@ object desugar {
             Nil
         }
       }
-      val classMods = if mods.is(Given) then mods &~ Given | Synthetic else mods
+      val classMods =
+        if mods.is(Given) then mods &~ Given | Synthetic | GivenClass
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala:813]: Instead of dropping `Given` and adding a new `GivenClass`, what prevents us from keeping `Given`?

File: tests/semanticdb/expect/InventedNames.expect.scala [hunk 3]
```diff
@@ -22,10 +22,10 @@ given `* */*<-givens::InventedNames$package.`* *`.*/`: Long/*->scala::Long#*/ =
 given X with
 /*<-givens::InventedNames$package.given_X.*//*->givens::X#*/  def doX/*<-givens::InventedNames$package.given_X.doX().*/ = 7
 
-given (using X/*->givens::X#*/): Y/*->givens::Y#*/ with
+/*->givens::InventedNames$package.given_Y().(x$1)*/given /*<-givens::InventedNames$package.given_Y().*/(using X/*->givens::X#*//*<-givens::InventedNames$package.given_Y().(x$1)*/): Y/*->givens::Y#*/ with
```
Comment (bishabosha) [inline tests/semanticdb/expect/InventedNames.expect.scala:25]: to fix the usage occurrence of `...given_Y().(x$1)` we need to avoid looking in the rhs of a `Given` method that returns a class with `GivenClass` flag.

The definition occurrences of `...given_Y().` and `...given_Y().(x$1)` are welcomed though.

File: tests/semanticdb/expect/InventedNames.expect.scala [hunk 4]
```diff
@@ -22,10 +22,10 @@ given `* */*<-givens::InventedNames$package.`* *`.*/`: Long/*->scala::Long#*/ =
...
   def doY/*<-givens::InventedNames$package.given_Y#doY().*/ = "7"
 
-given [T/*<-givens::InventedNames$package.given_Z_T#[T]*/]: Z/*->givens::Z#*/[T/*->givens::InventedNames$package.given_Z_T#[T]*/] with
+/*->givens::InventedNames$package.given_Z_T().[T]*/given /*<-givens::InventedNames$package.given_Z_T().*/[T/*<-givens::InventedNames$package.given_Z_T#[T]*//*<-givens::InventedNames$package.given_Z_T().[T]*/]: Z/*->givens::Z#*/[T/*->givens::InventedNames$package.given_Z_T#[T]*/] with
```
Comment (bishabosha) [inline tests/semanticdb/expect/InventedNames.expect.scala:28]: we probably want to substitute any occurrence for `...given_Z_T#[T]` with `...given_Z_T().[T]`, and the span of `...given_Z_T().[T]` should be where `...given_Z_T#[T]` is

File: tests/semanticdb/expect/Classes.expect.scala (hunk 1 above)
Comment (bishabosha) [reply tests/semanticdb/expect/Classes.expect.scala:14]: this can be solved by adding to `excludeDef` and `excludeChildren` a rule for implicit class conversions:
```scala
extension (sym: Symbol) // addition within `Scala3.SymbolOps` given
  def isImplicitClassConvertor(using Context) =
    sym.isAllOf(Implicit|Method) && sym.isCoDefinedGiven(sym.info.finalResultType.classSymbol)

...

private def excludeDef(sym: Symbol)(using Context): Boolean =
  !sym.exists
  || ...
  || sym.isImplicitClassConvertor
  || ...
  || excludeDefOrUse(sym)

...

private def excludeChildren(sym: Symbol)(using Context): Boolean =
  !sym.exists
  || ...
  || sym.isImplicitClassConvertor
```

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 5]
```diff
@@ -783,7 +783,7 @@ object desugar {
         DefDef(
           className.toTermName, joinParams(constrTparams, defParamss),
           classTypeRef, creatorExpr)
-          .withMods(companionMods | mods.flags.toTermFlags & GivenOrImplicit | Synthetic | Final)
+          .withMods(companionMods | mods.flags.toTermFlags & GivenOrImplicit | Final)
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala:786]: could an alternative be that in `Namer` we also look up specifically `isAllOf(Synthetic|Given, butNot = Private)` when looking up wildcard exports? then there is no need for TASTy changes.

**Review by bishabosha — DISMISSED** [review]
I think this can be done without the need for new flags and TASTy changes, unless there is a good reason for the new flag

File: tests/semanticdb/expect/InventedNames.expect.scala (hunk 3 above)
Comment (tgodzik) [reply tests/semanticdb/expect/InventedNames.expect.scala:25]: > The definition occurrences of ...given_Y(). and ...given_Y().(x$1) are welcomed though.

There are no actual occurrences in the source code, so in reality they are not useful for us.

Comment (bishabosha) [reply tests/semanticdb/expect/InventedNames.expect.scala:25]: ok so in this case there should be no occurrence for the definition for `given X with { ... }` either

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 5 above)
Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala:786]: se PR #12999

File: tests/semanticdb/expect/InventedNames.expect.scala (hunk 3 above)
Comment (tgodzik) [reply tests/semanticdb/expect/InventedNames.expect.scala:25]: yes, the previous state was exactly what we need.

File: tests/semanticdb/expect/Classes.expect.scala (hunk 1 above)
Comment (odersky) [reply tests/semanticdb/expect/Classes.expect.scala:14]: @bishabosha Can you take this over and do the exclude changes please?

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 5 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala:786]: That does not solve it since it still keeps the Synthetic flag on the given or implicit def. But I think I have it figured out now.

**Comment by bishabosha** [comment]
I think this should not be merged without a minor TASTy bump, it breaks the assumption that a type can not have Given or Implicit flag.

**Comment by odersky** [comment]
> I think this should not be merged without a minor TASTy bump, it breaks the assumption that a type can not have Given or Implicit flag. Unless we say that implementations should expect any possible flag

I think flags are sufficiently underspecified that we don't need to worry about this. The TreeUnpickler works unchanged for old as well as new version. That should be our test for version bumps. If an old TreeUnpickler works with the new format, we are good.

**Comment by bishabosha** [comment]
At present the Scala 2 TASTy reader scans flags not compatible with Scala 2 to see if it needs to special case certain definitions, (such as `Invisible`) - so currently it explicitly checks for certain "redundant" flags it can safely ignore, such as `Opaque` on a class. So if a definition suddenly had a new flag that is not known to be safe to ignore, then the user will get an error.  Perhaps this approach was wrong.

TLDR: this will break Scala 2 TASTy reader for previously accepted code

**Comment by odersky** [comment]
OK, we have this under next minor release anyway.

File: tests/semanticdb/expect/Classes.expect.scala (hunk 1 above)
Comment (tanishiking) [reply tests/semanticdb/expect/Classes.expect.scala]: It looks like it has been fixed in https://github.com/lampepfl/dotty/pull/12885 (specifically https://github.com/lampepfl/dotty/pull/12885/commits/fef5be020c9e7c02446772140910b9e6a03aeaa8 ), those occurrences will disappear once we rebase onto master.

File: tests/semanticdb/expect/InventedNames.expect.scala (hunk 3 above)
Comment (tanishiking) [reply tests/semanticdb/expect/InventedNames.expect.scala:25]: Same with https://github.com/lampepfl/dotty/pull/12979#discussion_r689259730

I tried rebasing onto master to re-generated expected files, and now I see: 

```scala
given (using X/*->givens::X#*/): Y/*->givens::Y#*/ with
  def doY/*<-givens::InventedNames$package.given_Y#doY().*/ = "7"

given [T/*<-givens::InventedNames$package.given_Z_T#[T]*//*<-givens::InventedNames$package.given_Z_T().[T]*/]: Z/*->givens::Z#*/[T/*->givens::InventedNames$package.given_Z_T#[T]*/] with
  def doZ/*<-givens::InventedNames$package.given_Z_T#doZ().*/: List/*->scala::package.List#*/[T/*->givens::InventedNames$package.given_Z_T#[T]*/] = Nil/*->scala::package.Nil.*/
```

**Comment by bishabosha** [comment]
this needs a rebase

**Comment by odersky** [comment]
@bishabosha Can you check that the semanticdb expect files are now correct? Thanks!

**Comment by bishabosha** [comment]
@odersky please may you check the addition I made, otherwise I think the semanticdb changes are good

File: compiler/src/dotty/tools/dotc/transform/SymUtils.scala [hunk 6]
```diff
@@ -100,6 +100,17 @@ object SymUtils:
         case _ =>
           false
 
+    def isGivenInstanceSummoner(using Context): Boolean =
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/SymUtils.scala:103]: The code looks good but I am not sure why given instance summoners should not be reported in semanticdb. Can you explain?

Comment (tgodzik) [reply compiler/src/dotty/tools/dotc/transform/SymUtils.scala:103]: Nothing should be reported in semanticdb that doesn't have a real source code position. It can be added to synthetic section, but that will be done later on.

Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/transform/SymUtils.scala:103]: then again with a given instance I think the mental model is we are defining a term, not a class

File: tests/semanticdb/expect/InventedNames.expect.scala (hunk 3 above)
Comment (bishabosha) [reply tests/semanticdb/expect/InventedNames.expect.scala]: fixed by https://github.com/lampepfl/dotty/pull/12979/commits/ec3a21a0065799d4d9d2eb11452330ff5ee556ae

File: compiler/src/dotty/tools/dotc/transform/SymUtils.scala (hunk 6 above)
Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/transform/SymUtils.scala:103]: After discussion we decided to try and favour the given instance method over the structural given class

**Comment by bishabosha** [comment]
The problem with the structural given instance summoner method is that the return type is a `TypeTree` and so again its positions are lost

File: compiler/src/dotty/tools/dotc/transform/SymUtils.scala (hunk 6 above)
Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/transform/SymUtils.scala:103]: the issue we have is that
```scala
given [T]: Z[T] with {}
```
generates
```scala
final given def given_Z_T[T]: given_Z_T[T] = new given_Z_T[T]
given class given_Z_T[T] extends Z[T]
```
so it would be correct for occurrences in `: Z[T]` to reference the `T` in the *class* `given_Z_T`, so the only question is what should the definition of the type parameter `given [T]: ` belong to in semanticdb - we could say the method, but that is quite strange visually, as then you get the reference in the return type coming from a different definition

Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/transform/SymUtils.scala:104]: commit https://github.com/lampepfl/dotty/pull/12979/commits/7a01b0f52cff19c6a7fe28377a5e65a69ef3531d highlights this difference

**Comment by bishabosha** [comment]
@odersky I have updated so that in semanticdb the type parameter in the structural given is recorded as belonging to the summoner method

File: compiler/src/dotty/tools/dotc/transform/SymUtils.scala [hunk 7]
```diff
@@ -87,6 +87,34 @@ object SymUtils:
...
+      self.is(Implicit) && self.info.stripPoly.match
+        case mt @ MethodType(_ :: Nil) if !mt.isImplicitMethod =>
+          if self.isCoDefinedGiven(mt.finalResultType.typeSymbol)
+          then !directOnly
+          else !forImplicitClassOnly
+        case _ =>
+          false
+
+    /** Is this the method that summons a structural given instance? */
+    def isGivenInstanceSummoner(using Context): Boolean =
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/transform/SymUtils.scala:104]: This method isn't actually used now, but it seems potentially useful to keep?

File: compiler/src/dotty/tools/dotc/transform/SymUtils.scala (hunk 6 above)
Comment (tgodzik) [reply compiler/src/dotty/tools/dotc/transform/SymUtils.scala:104]: I would say `*<-givens::InventedNames$package.given_Z_T#[T]` looks better, because givens having the `with` cause look more akin do classes than methods, but on the other hand this is neither method nor class from the code point of view. We could introduce another way of showing givens (we are already discussing a new symbol construct for extension methods), until then in reality both approaches are ok to use here.
